lib/zclient.c

```diff
@@ -2285,12 +2286,26 @@ struct interface *zebra_interface_link_params_read(struct stream *s,
 		return NULL;
 	}
 
+	if (changed == NULL)
+		return NULL;
+
+	*changed = false;
+	struct if_link_params iflp_copy;
+
+	if (ifp->link_params == NULL)
+		*changed = true;
+
 	if ((iflp = if_link_params_get(ifp)) == NULL)
 		return NULL;
 
+	memcpy(&iflp_copy, iflp, sizeof(iflp_copy));
+
 	if (link_params_set_value(s, iflp) != 0)
 		goto stream_failure;
 
+	if (memcmp(&iflp_copy, iflp, sizeof(iflp_copy)))
+		*changed = true;
+
```

Comment:
Here, the goal is to check that there is really a modification in the link parameters. Can you point me the scenario when this function is called while link parameters are not modified? Does this happen when another interface parameter is modified e.g. address, bandwidth, speed, metric ... ? If yes, I think it will be better to change `zserv_encode_interface()` in zapi_msg.c file in order to not transmit link parameters if they have not change. I.e. lines 93 - 98 of zapi_msg.c to check that link parameters have change or not like you propose here. By doing this, we avoid sending an unnecessary too long message via zebra.

Reply:
One scenario that triggers sending a ZEBRA_INTERFACE_LINK_PARAMS message (when link-params are not changed is:

- enabling tcpdump

Here's the call to zebra_interface_up_update that is called when tcpdump is enabled:
https://github.com/FRRouting/frr/blob/master/zebra/if_netlink.c#L1708

Reply:
I> Here, the goal is to check that there is really a modification in the link parameters. Can you point me the scenario when this function is called while link parameters are not modified? Does this happen when another interface parameter is modified e.g. address, bandwidth, speed, metric ... ? If yes, I think it will be better to change `zserv_encode_interface()` in zapi_msg.c file in order to not transmit link parameters if they have not change. I.e. lines 93 - 98 of zapi_msg.c to check that link parameters have change or not like you propose here. By doing this, we avoid sending an unnecessary too long message via zebra.

@odd22 Thanks for the code pointer.

In order for zapi_msg.c to detect that link_params have not changed, I believe that 'struct interface *ifp' would need to store a copy of the link_params that was last sent to a client.  

If the ifp had a copy of link_params_sent, then zsend_interface_link_params() could check if the link_params changed (and only send if there is a change).

As an alternate possible solution, perhaps it is not necessary for zebra_interface_add_update() to call zsend_interface_link_params().  
Perhaps it is only necessary for zebra to call zsend_interface_link_params() when zebra/interface.c processes a link-params CLI command and calls zebra_interface_parameters_update().
https://github.com/FRRouting/frr/blob/master/zebra/redistribute.c#L501

Thanks, Karen

Reply:
I deeply investigate the `zebra/redistribute.c` file as well as `zebra/zapi_msg.c` and I discover that `zebra_interface_parameters_update()` is called in many wrong place. This function must be call only when link parameters are modified i..e by `zebra_interface_parameters_update()`. All other place where the function is called is wrong because there is a call to `zsend_interface_add()` or `zsend_interface_update()` just before. However, both function call `zserv_encode_interface()` which `zebra_interface_link_params_write()` to add link parameters to the message is Link Param is enable. Thus, I think it is not necessary to send twice the link parameters in particular because there have not been modified.

A simple way to verify that link params have been modified will be to use the `lp_status` in `struct if_link_params`. Indeed, not all bits of the status are used. Thus, we could add a "change bit" (say bit 31) to indicate that link params have been changed. This bit will be set when CLI change a link param and reset when link params are sent through the `zebra_interface_parameters_update()`. This way, it is not necessary to keep a copy of link params and we could keep call to `zebra_interface_parameters_update()` in many place if is justify. Link params will be sent only if they have been modified.